qualtran/bloqs/rotations/zpow_via_phase_gradient.py

```diff
@@ -0,0 +1,121 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import cast
+
+import sympy
+from attrs import frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BloqBuilder,
+    DecomposeTypeError,
+    QBit,
+    QFxp,
+    Signature,
+    Soquet,
+    SoquetT,
+)
+from qualtran.bloqs.arithmetic import XorK
+from qualtran.bloqs.rotations.phase_gradient import AddIntoPhaseGrad
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.resource_counting.generalizers import ignore_alloc_free
+from qualtran.symbolics import ceil, is_symbolic, log2, pi, SymbolicFloat, SymbolicInt
+
+
+@frozen
+class ZPowConstViaPhaseGradient(Bloq):
+    r"""Apply an $ZPow(t)$ using a phase gradient state."""
+    exponent: SymbolicFloat
+    phase_grad_bitsize: SymbolicInt
+
+    @property
+    def signature(self) -> 'Signature':
+        return Signature.build_from_dtypes(q=QBit(), phase_grad=self.phase_grad_dtype)
+
+    @classmethod
+    def from_precision(
+        cls, exponent: SymbolicFloat, *, eps: SymbolicFloat
+    ) -> 'ZPowConstViaPhaseGradient':
+        b_grad = ceil(log2(2 * pi(eps) / eps))
+        return cls(exponent, b_grad)
+
+    @cached_property
+    def phase_grad_dtype(self) -> QFxp:
+        return QFxp(self.phase_grad_bitsize, self.phase_grad_bitsize)
+
+    @property
+    def _load_bloq(self) -> XorK:
+        if is_symbolic(self.exponent) or is_symbolic(self.phase_grad_bitsize):
+            return XorK(self.phase_grad_dtype, cast(sympy.Expr, self.exponent / 2))
+
+        k_int = self.phase_grad_dtype.to_fixed_width_int(self.exponent / 2)
+        return XorK(self.phase_grad_dtype, k_int)
+
+    def build_composite_bloq(
+        self, bb: BloqBuilder, q: Soquet, phase_grad: Soquet
+    ) -> dict[str, SoquetT]:
+        if is_symbolic(self.exponent):
+            raise DecomposeTypeError(f"cannot decompose {self} with symbolic {self.exponent=}")
+
+        # TODO this could use `AddK` once it supports QFxp.
```

Comment:
`AddK` would have the same complexity and it would still load the angle in an ancilla and then do quantum-quantum addition; correct? 

Either link the TODO to a github issue or remove. (I think it's fine to remove).

Reply:
Yes that is correct. I removed it.